### PR TITLE
[#1164] Document and improve message upsert endpoint

### LINKED ISSUE #1164: Document and improve `messages.upsert`
Sending fields that are not string values silently fails, but there should be an error. We also took decisions around implicitly mapping fields ending with `count` to integers and fields ending with `content` to objects, so we should document that.

Also we will fix the glossary:

Currently it says: Metadata is data attached to a conversation consisting of a set of Key/Value pairs. A key can use the dot notation to represent namespaces. which is not accurate anymore as we now can attach metadata to more things.

### PR DESCRIPTION
**Pull request by chrismatix**
resolves #1164

### CODE REVIEW

File: docs/docs/concepts/metadata.md [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+title: Metadata
+sidebar_label: Metadata
+---
+
+We address the main use case of a messaging platform with conversations, channels and messages. However, this
+is not enough for users to address the wide variety of messaging workflows that exist such as automating and
+suggesting replies, attaching contact information etc.
+
+For this reason we introduced an extension to the data model that we call metadata. Metadata in the context
+of the Airy Core platform is an optional document that gets attached to a subject, which consists of a
```
Comment (lucapette) [inline docs/docs/concepts/metadata.md:11]:
```suggestion
of the Airy Core platform is an optional document you can attach to a subject, which consists of a
```

File: docs/docs/concepts/metadata.md [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+This metadata then gets exposed on our [HTTP APIs](/api/endpoints/introduction.md), see for instance the 
+`conversation.metadata` field, and it can be streamed via the [websocket](/api/websocket.md) and [webhook](/api/webhook.md). 
+ 
+
+### Subject
+
+A metadata subject is comprised of a namespace such as `conversation` and a namespace identifier, which is
+the identifier that uniquely identifies the entity the metadata is bound to within the namespace.
+
+So the subject of a conversation with id `123` would be `conversation:123`.
```
Comment (lucapette) [inline docs/docs/concepts/metadata.md:23]:
```suggestion
Therefore the subject of a conversation with id `123` would be `conversation:123`.
```

File: docs/docs/concepts/metadata.md [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+{
+  "contact": {
+    "first_name": "Grace",
+    "last_name": "Hopper"
+  },
+  "state":  "open"
+}
+```
+
+Currently, this design introduces two limitations on the structure of metadata:
```
Comment (lucapette) [inline docs/docs/concepts/metadata.md:53]: I wonder if this should be a paragraph, like  `## Limitations` or something like that (no strong opinion but I thought about it while scanning the docs so I thought worth sharing)

File: docs/docs/concepts/metadata.md [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+
+We address the main use case of a messaging platform with conversations, channels and messages. However, this
+is not enough for users to address the wide variety of messaging workflows that exist such as automating and
+suggesting replies, attaching contact information etc.
+
+For this reason we introduced an extension to the data model that we call metadata. Metadata in the context
+of the Airy Core platform is an optional document that gets attached to a subject, which consists of a
+namespace i.e. "conversation" and an identifier within that namespace i.e. the conversation id.
+
+This metadata then gets exposed on our [HTTP APIs](/api/endpoints/introduction.md), see for instance the
```
Comment (lucapette) [inline docs/docs/concepts/metadata.md:14]:
```suggestion
Our [HTTP APIs](/api/endpoints/introduction.md) exposes this metadata, see for instance the 
```

File: docs/docs/concepts/metadata.md [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+JSON dot notation to reflect nested data. Take for instance the following series of metadata records
+for a conversation:
+
+| Key                | Value  |
+|--------------------|--------|
+| contact.first_name | Grace  |
+| contact.last_name  | Hopper |
+| state              | open   |
+
+For the API payloads this gets assembled to the following metadata document:
```
Comment (lucapette) [inline docs/docs/concepts/metadata.md:41]:
```suggestion
The API payloads expose this list as the following metadata document:
```

**Review by lucapette — APPROVED** [review]
Amazing work 👏 


I tried to eliminate passive voice everywhere with the suggestions (nothing major though)

File: docs/docs/concepts/metadata.md (hunk 3 above)
Comment (chrismatix) [reply docs/docs/concepts/metadata.md:53]: True! I will add it.

**Review by lucapette — DISMISSED** [review]
:shipit:

**Review by lucapette — APPROVED** [review]

**Pull request merged**